[clang-repl] Improve error message on failed undos

aadanen opened the pull request:
Updated error message logic for undo function. Throws different errors for the case of there being nothing to undo, and for the case of requesting more undos than there are operations to undo.

Fixes https://github.com/llvm/llvm-project/issues/143668

aadanen committed:
clarified %undo error msg in clang-repl

used grep -r "Too many undos" to locate all instances where the string
was mentioned. Changed the string to the suggestion in issue 143668

aadanen committed:
Merge branch 'main' of github.com:aadanen/llvm-project into issue143668

aadanen committed:
Merge branch 'llvm:main' into issue143668

DavidSpickett commented on clang/unittests/Interpreter/InterpreterTest.cpp:
```diff
@@ -158,12 +158,12 @@ TEST_F(InterpreterTest, UndoCommand) {
   EXPECT_TRUE(!!Err2);
   auto Err3 = Interp->Undo(2);
-  EXPECT_EQ("Operation failed. Too many undos",
+  EXPECT_EQ("Operation failed. No input left to undo",
```
The complication here is that although the clang-repl `%undo` only undoes one input, the internal method can actually undo more than that. This test for example is trying to undo 2 things.

So the code needs to be a bit more intelligent. You have two situations:
* You have nothing left to undo.
* You have some things left to undo but *want* to undo more than that.

In both cases you error because you don't want to partially undo something, but the message should be specific. In the same order as above, I suggest:
* "No input left to undo"
* "Want to undo X inputs, only have Y" (with the message formatted with the correct numbers)

For formatting the error, we have some utilities to do that, I will find those for you.

Side note: "inputs" might not be the right noun, but stick with it for now and we can ask a clang-repl maintainer whether it's ok.

DavidSpickett replied:
I guess that the internal method can undo more than one input because you might be piping many lines at once into clang-repl? Anyway it's there so we should assume it is there for a good reason.

DavidSpickett commented:
https://github.com/llvm/llvm-project/pull/149396#issuecomment-3088555387 is one of our automated checks, please follow its guidance. The reason we do this is so we can attribute changes for reasons around the licensing of contributions, there's more info in those links. It can be any valid address, you don't have to use a "personal" one, for instance I use my employer's email address (that and they require that I do, but beside the point).

aadanen committed:
implemented contextual error msg

differentiate between the case of 0 inputs to undo and the case when
there are X inputs and Y requested undos but Y > X

aadanen commented:
I put in inputs vs undo requests logic but I think the unit test might need to be edited as well

aadanen committed:
Slight edit to error msg. Updated unit test.

aadanen committed:
lil type made me fail the tests

aadanen committed:
Merge branch 'main' into issue143668

aadanen commented:
Okay I think that does it :)

DavidSpickett commented:
Looks great! CI should confirm the test is correct, and I added the clang-repl maintainer on review in case they have anything to say.

DavidSpickett commented:
> I am able to build clang, compile the project, and I am currently running the "ninja check-clang-unit" tests. Although it does take an incredibly long time ( 45 minutes to an hour) but that's off topic

llvm/clang is not a light build at the best of times, but there are some things you can try - https://llvm.org/docs/GettingStarted.html#common-problems

If you do pick up another issue, I can give you some tailored suggestions for the specific task. For instance here, "check-clang" is runs all clang checks but I happened to know that the test you're changing is what we call a "unit test" and has it's own smaller target.

> some of the commits in this PR are me updating the branch with changes to the main branch. I think its okay because this whole thing will get squashed at the end but just checking

That's right. In fact I think those "merge" commits are preferred during review, versus rebasing, because GitHub likes to hide review state if you do the latter. It's also fine to not update the PR branch until review has concluded.

DavidSpickett commented:
You have formatting to fix, I suggest you just copy and paste the diff in this one time.

Folks will have different ways to setup clang-format, I use the script https://clang.llvm.org/docs/ClangFormat.html#script-for-patch-reformatting and manually run it (because I am too lazy to integrate it somewhere :) ).

I also add "black" for python formatting, into one giant command:
```
git diff -U0 --no-color --relative HEAD^ | ./clang/tools/clang-format/clang-format-diff.py -p1 -i -binary /home/david.spickett/build-llvm-aarch64/bin/clang-format && darker --revision HEAD~1 $(git diff-tree --no-commit-id --name-only HEAD -r)
```

If you're using an editor, do look into applying the format on save, it's much more convenient.

vgvassilev reviewed (approved):
Lgtm! Thank you @aadanen!

DavidSpickett reviewed (approved):
LGTM

@aadanen you won't be able to merge this yourself but I can once you've fixed the formatting.

aadanen committed:
formatting

aadanen committed:
Merge branch 'issue143668' of github.com:aadanen/llvm-project into issue143668

aadanen committed:
Merge branch 'main' into issue143668
